### PR TITLE
Implemented generic os event driven mechanism

### PR DESCRIPTION
**Pull request by vidhutsingh**
We wanted to move the changes from https://github.com/hpe-storage/csi-driver/pull/145 PR to common-host-util as it is a generic mechanism. All the details are captured in the original PR .

### CODE REVIEW

File: util/watcher.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	}()
+
+	return watch, nil
+}
+
+//AddWatchList list of files /and directories to watch
+func (w *FileWatch) AddWatchList(files []string) error {
+	log.Trace(">>>>> AddWatchList")
+	defer log.Trace("<<<<< AddWatchList")
+	if len(files) == 0 {
```
Comment (rgcostea) [inline util/watcher.go:73]: nit: new line

File: util/watcher.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+		return fmt.Errorf("Empty watch list is not supported, there should be at least one file to watch")
+	}
+	for _, fPath := range files {
```
Comment (rgcostea) [inline util/watcher.go:76]: nit: new line

File: util/watcher.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
```
Comment (rgcostea) [inline util/watcher.go:77]: delete this line

File: util/watcher.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		err := w.watchList.Add(fPath)
+		if err != nil {
+			log.Warnf("Failed to add [%s] file to watch list, err %s :", fPath, err.Error())
+		} else {
+			log.Tracef("Successfully added [%s] file to watch list", fPath)
+		}
+
```
Comment (rgcostea) [inline util/watcher.go:84]: delete this line

File: util/watcher.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// StartWatcher triggers watcher until os sig interrupt. This will run anonymous fn forever.
+func (w *FileWatch) StartWatcher() {
+	log.Trace(">>>>> StartWatcher")
+	defer log.Trace("<<<<< StartWatcher")
+	pid := os.Getpid()
+	log.Tracef("Watcher [%d PID] successful started", pid)
+	// forever
+	for {
```
Comment (rgcostea) [inline util/watcher.go:96]: new line here

File: util/watcher.go (hunk 3 above)
Comment (vidhutsingh) [reply util/watcher.go:77]: Done

File: util/watcher.go (hunk 4 above)
Comment (vidhutsingh) [reply util/watcher.go]: Done

File: util/watcher.go (hunk 5 above)
Comment (vidhutsingh) [reply util/watcher.go:96]: Done

**Comment by vidhutsingh** [comment]
thanks @rgcostea for code review.

File: util/watcher.go [hunk 6]
```diff
@@ -91,7 +92,17 @@ func (w *FileWatch) StartWatcher() {
 	log.Trace(">>>>> StartWatcher")
 	defer log.Trace("<<<<< StartWatcher")
 	pid := os.Getpid()
-	log.Tracef("Watcher [%d PID] successful started", pid)
+	log.Tracef("Watcher [%d PID] is successfully started", pid)
+
+	// Control the ticker interval, dont want to frequently wakeup
+	// watcher as it is only needed when there is event notification. So if there is
+	// event notification, ticker is set to wake up every one miniute otherwise sleep
```
Comment (rgcostea) [inline util/watcher.go:99]: minute

File: util/watcher.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	"sync"
+	"syscall"
+	"time"
+
+	notify "github.com/fsnotify/fsnotify"
+	log "github.com/hpe-storage/common-host-libs/logger"
+)
+const tickerDefaultDelay = 60 // Min
+// FileWatch contains watcher attributes.
+type FileWatch struct {
```
Comment (rgcostea) [inline util/watcher.go:18]: nit: Add some new lines here.

**Pull request merged**